`LocalTransaction` test: sort transactions when nominating

hewison-chris opened the pull request:
This test still sometimes fails when waiting for the block to be externalized. 
The transactions should be sorted when nominating.

Geod24 commented:
> source/agora/test/LocalTransactions.d(82,46): Error: undefined identifier `GenesisBlock`

hewison-chris commented:
> > source/agora/test/LocalTransactions.d(82,46): Error: undefined identifier `GenesisBlock`

🤦

Geod24 commented on source/agora/test/Base.d:
```diff
@@ -1972,7 +1972,7 @@ public struct TestConf
     /// max failed requests before a node is banned
     /// Matches the eponymous field in the `banman` section.
-    public size_t max_failed_requests = 100;
+    public size_t max_failed_requests = 10_000; // default to not banning in tests
```
Note that banning is only done on non-validator, so I doubt that's it.

hewison-chris replied:
Some tests have `FullNodes` too. But it probably will not solve much.

linked0 replied:
I saw some nodes are banned in unit tests sometimes. We should check why some nodes are banned just in test cases, which is unexpected. I'm not sure we need to create a new issue about that though.

Geod24 replied:
The nodes do ban themselves, so you will see a message saying "node-7" banned, but it will be emitted from node-7.

hewison-chris commented:
It passed required. I have set to re-run to see if it passes again.

kchulj commented:
> It passed required. I have set to re-run to see if it passes again.

How is it looking? It seems to pass.

hewison-chris committed:
Trivial whitespace removal

linked0 commented on source/agora/test/ValidatorRecurringEnrollment.d:
```diff
@@ -241,8 +241,8 @@ unittest
     // Make 2 nodes sleep for longer than test can possibly run
-    node_4.ctrl.sleep(180.seconds, true);
-    node_5.ctrl.sleep(180.seconds, true);
+    node_4.ctrl.sleep(1.hours, true);
```
So big time. Is there any reason for `1.hours` instead of a dozen of minutes?

hewison-chris replied:
Not really, but it will be cancelled when `sleep(0)` is called later so it could be `1.year` even.

linked0 reviewed (approved):
LGTM. I hope this will prevent the spurious test error forever. :)

hewison-chris commented:
> LGTM. I hope this will prevent the spurious test error forever. :)

I fear this is not the end of these problems as it is still sensitive to changes.

hewison-chris commented:
Ubuntu master failed in last run with 30 min timeout so re-running :-(

omerfirmak commented:
From https://github.com/bosagora/agora/pull/2404/checks?check_run_id=3408574500

They are al nominating same TXs in different order and failing to reach consensus

```
2021-08-24 07:51:52,283 Info [agora.consensus.protocol.Nominator] - Nominating { tx_set: [0x4a26...dc5c, 0x78ea...5b93, 0x779d...3f94, 0x34f8...a1b8, 0x68c3...927d, 0x7dd1...37cd], enrolls: [], missing_validators: [], time_offset: 600 } at 1609459800
```
```
2021-08-24 07:51:10,668 Info [agora.consensus.protocol.Nominator] - Nominating { tx_set: [0x7dd1...37cd, 0x78ea...5b93, 0x4a26...dc5c, 0x779d...3f94, 0x34f8...a1b8, 0x68c3...927d], enrolls: [], missing_validators: [], time_offset: 600 } at 1609459800
```
```
2021-08-24 07:50:52,146 Info [agora.consensus.protocol.Nominator] - Nominating { tx_set: [0x34f8...a1b8, 0x78ea...5b93, 0x4a26...dc5c, 0x779d...3f94, 0x7dd1...37cd, 0x68c3...927d], enrolls: [], missing_validators: [], time_offset: 600 } at 1609459800
```
```
2021-08-24 07:50:53,191 Info [agora.consensus.protocol.Nominator] - Nominating { tx_set: [0x68c3...927d, 0x78ea...5b93, 0x4a26...dc5c, 0x779d...3f94, 0x34f8...a1b8, 0x7dd1...37cd], enrolls: [], missing_validators: [], time_offset: 600 } at 1609459800
```
```
2021-08-24 07:50:49,210 Info [agora.consensus.protocol.Nominator] - Nominating { tx_set: [0x779d...3f94, 0x78ea...5b93, 0x4a26...dc5c, 0x34f8...a1b8, 0x68c3...927d, 0x7dd1...37cd], enrolls: [], missing_validators: [], time_offset: 600 } at 1609459800
```
```
2021-08-24 07:50:47,168 Info [agora.consensus.protocol.Nominator] - Nominating { tx_set: [0x78ea...5b93, 0x4a26...dc5c, 0x779d...3f94, 0x34f8...a1b8, 0x7dd1...37cd, 0x68c3...927d], enrolls: [], missing_validators: [], time_offset: 600 } at 1609459800
```

Would sorted transactions help?

hewison-chris committed:
Always sort the `transactions` in `getCandidateTransactions`

The transactions were only being sorted if the number of transactions
was `MaxTransactionsPerBlock` which is currently 1000. If it is a payout
block then the Coinbase will still be added as the last.

kchulj reviewed (approved):
LGTM

Geod24 commented:
```
Unittesting agora.test.GenesisBlock (took 2 secs, 361 ms, 99 μs, and 6 hnsecs)
Unittesting agora.test.Flash#### FATAL ERROR: Connection with peer closed
This node was started at source/agora/test/Flash.d:399
This most likely means that the node crashed due to an uncaught exception
If not, please file a bug at https://github.com/Geod24/localrest/
./ci/run.sh: line 21: 13495 Abort trap: 6           ./build/agora-unittests
Full error: object.Exception@submodules/localrest/source/geod24/LocalRest.d-mixin-1179(1190): Connection with peer closed
+ ./build/agora-unittests
----------------
/Users/runner/work/agora/agora/submodules/localrest/source/geod24/LocalRest.d-mixin-1179:1190 @trusted geod24.LocalRest.Response geod24.LocalRest.RemoteAPI!(agora.test.Flash.TestFlashAPI, agora.test.Base.Serializer).RemoteAPI.requestSettleSig(agora.crypto.Key.PublicKey, agora.crypto.Key.PublicKey, geod24.bitblob.BitBlob!(64uL).BitBlob, uint).__lambda6() [0x107763ad1]
/Users/runner/work/agora/agora/submodules/localrest/source/geod24/LocalRest.d-mixin-1179:1185 @safe agora.flash.Types.Result!(agora.crypto.Schnorr.Signature).Result geod24.LocalRest.RemoteAPI!(agora.test.Flash.TestFlashAPI, agora.test.Base.Serializer).RemoteAPI.requestSettleSig(agora.crypto.Key.PublicKey, agora.crypto.Key.PublicKey, geod24.bitblob.BitBlob!(64uL).BitBlob, uint) [0x10770001f]
/Users/runner/work/agora/agora/submodules/localrest/source/geod24/LocalRest.d-mixin-1179:1180 _DThn16_6geod249LocalRest__T9RemoteAPITC5agora4test5Flash12TestFlashAPITSQBhQBe4Base10SerializerZQCr16requestSettleSigMFNfSQDf6crypto3Key9PublicKeyQzSQFm7bitblob__T7BitBlobVmi64ZQokZSQFn5flash5Types__T6ResultTSQGoQDj7Schnorr9SignatureZQBi [0x107765a28]
/Users/runner/work/agora/agora/source/agora/flash/UpdateSigner.d:335 agora.flash.Types.Result!(agora.flash.Types.UpdatePair).Result agora.flash.UpdateSigner.UpdateSigner.collectSignatures(agora.flash.api.FlashAPI.FlashAPI, uint, agora.consensus.data.Transaction.Output[], agora.flash.Types.PrivateNonce, agora.flash.Types.PublicNonce, geod24.bitblob.BitBlob!(64uL).BitBlob, agora.consensus.data.UTXO.UTXO) [0x107842998]
/Users/runner/work/agora/agora/source/agora/flash/Channel.d:994 void agora.flash.Channel.Channel.checkProposeUpdate() [0x1073b6db2]
/Users/runner/work/agora/agora/source/agora/flash/Channel.d:602 void agora.flash.Channel.Channel.eventLoop() [0x1078697d7]
/Users/runner/work/agora/agora/source/agora/flash/Channel.d:533 void agora.flash.Channel.Channel.start(in ref agora.flash.Types.PrivateNonce, in ref agora.flash.Types.PublicNonce) [0x107867a3b]
/Users/runner/work/agora/agora/source/agora/flash/Channel.d:191 void agora.flash.Channel.Channel.__ctor(agora.flash.Config.FlashConfig, agora.flash.Config.ChannelConfig, in ref agora.crypto.Key.KeyPair, agora.flash.Types.PrivateNonce, agora.flash.Types.PublicNonce, agora.flash.api.FlashAPI.FlashAPI, agora.script.Engine.Engine, agora.common.Task.ITaskManager, void delegate(agora.consensus.data.Transaction.Transaction), void delegate(in ref agora.crypto.Key.PublicKey, in ref geod24.bitblob.BitBlob!(64uL).BitBlob, in ref geod24.bitblob.BitBlob!(64uL).BitBlob, in agora.common.Amount.Amount, in agora.common.Types.Height, in ref agora.flash.OnionPacket.OnionPacket), void delegate(agora.crypto.Key.PublicKey, geod24.bitblob.BitBlob!(64uL).BitBlob, agora.flash.Types.ChannelState, agora.flash.ErrorCode.ErrorCode) @safe, void delegate(agora.crypto.Key.PublicKey, geod24.bitblob.BitBlob!(64uL).BitBlob, geod24.bitblob.BitBlob!(64uL).BitBlob, agora.flash.ErrorCode.ErrorCode) @safe, void delegate(agora.crypto.Key.PublicKey, ...
/Users/runner/hostedtoolcache/dc/ldc2-1.26.0/x64/ldc2-1.26.0-osx-x86_64/bin/../import/core/internal/lifetime.d:56 void geod24.LocalRest.Timer.run() [0x107e7e9fb]
context.d:46 void core.thread.context.Callable.opCall() [0x1083e867e]
fiber.d:1129 void core.thread.fiber.Fiber.run() [0x1083e880d]
fiber.d:186 fiber_entryPoint [0x1083e8742]
```

FYI @omerfirmak

Geod24 commented:
```
0   agora-unittests                     0x00000001103f6ff8 _D4core7runtime9backtraceFNbPPviZi + 88
1   agora-unittests                     0x00000001103f6f6c _D4core7runtime18runModuleUnitTestsUZ19unittestSegvHandlerUNbiPSQCk3sys5posix6signal9siginfo_tPvZv + 76
2   libsystem_platform.dylib            0x00007fff7356b5fd _sigtramp + 29
3   agora-unittests                     0x000000010f818d4c _D5agora4node9ValidatorQk17regenerateQuorumsMFNbNfSQBx6common5Types6HeightZv + 60
4   agora-unittests                     0x000000010f6bf6e9 _D5agora4node9ValidatorQk15onAcceptedBlockMFNfIKSQBv9consensus4data5BlockQgbZv + 281
5   agora-unittests                     0x000000010f6f3d5c _D5agora9consensus6LedgerQh11acceptBlockMFNfIKSQBtQBq4data5BlockQgZAya + 636
6   agora-unittests                     0x000000010f82cce0 _D5agora4node8FullNodeQj9addBlocksMFNfAxSQBn9consensus4data5BlockQgAxSQCqQBdQw12PreImageInfoQoZSQDq6common5Types6Height + 256
7   agora-unittests                     0x000000010f45f18f _D5agora7network7Manager14NetworkManager13getBlocksFromMFNbSQCg6common5Types6HeightMDFNfAxSQDl9consensus4data5BlockQgAxSQEoQBdQw12PreImageInfoQoZQDiZv + 1087
8   agora-unittests                     0x000000010f6bd5a2 _D5agora4node8FullNodeQj11catchupTaskMFNbZv + 178
9   agora-unittests                     0x000000010fe909fc _D6geod249LocalRest5Timer3runMFZv + 92
10  agora-unittests                     0x00000001103fa67f _D4core6thread7context8Callable6opCallMFZv + 63
11  agora-unittests                     0x00000001103fa80e _D4core6thread5fiber5Fiber3runMFZv + 30
12  agora-unittests                     0x00000001103fa743 fiber_entryPoint + 147
./ci/run.sh: line 21: 31514 Segmentation fault: 11  ./build/agora-unittests
Unittesting agora.test.MissingPreImageDetection
```
